### PR TITLE
Niak 55 Spring security

### PR DESCRIPTION
**Pull request by Sewerinius**
Added Spring security to the backend.
Secured all the endpoints with 4 roles: 
"ROLE_ANON", "ROLE_STANDARD", "ROLE_SUPERVISOR", "ROLE_ADMINISTRATOR"
All endpoints are forbidden by default
There is an "/login" endpoint which you can use to log in
- The first user needs to be added to the database manually (you can use something like "{noop}pass" for password)

### CODE REVIEW

**Comment by Sewerinius** [comment]
Now that I did something on the backend I have something to say about all those Request and Response classes.
As somebody correctly put them in a rest package: they are part of the rest logic so they shouldn't inside of services (which are supposed to do business logic). 
I think we should only use them to encapsulate objects inside of the Controller classes.

**Comment by enthyp** [comment]
Some conflicts appeared, I'll try to check this out tomorrow

File: rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/SystemUserController.java [hunk 1]
```diff
@@ -53,27 +61,41 @@ public GetSystemUsersResponse getSystemUsersByFirstNameAndLastName(@RequestParam
...
     }
+
+    @Secured({"ROLE_ANON", "ROLE_STANDARD", "ROLE_SUPERVISOR", "ROLE_ADMINISTRATOR"})
+    @GetMapping(path = "me")
+    @ResponseBody
+    @ResponseStatus(value = HttpStatus.OK)
+    public Object getMe() {
+        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+        System.out.println(auth);
+
```
Comment (enthyp) [inline rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/SystemUserController.java]: I think Sout should be gone

File: rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/RecruitmentController.java [hunk 2]
```diff
@@ -20,24 +21,27 @@
 @RestController
 @RequestMapping("recruitment")
 @AllArgsConstructor(onConstructor = @__(@Autowired))
-public final class RecruitmentController {
+public class RecruitmentController {
 
     private final RecruitmentService recruitmentService;
 
+    @Secured({"ROLE_STANDARD", "ROLE_SUPERVISOR", "ROLE_ADMINISTRATOR"})
     @PostMapping
```
Comment (enthyp) [inline rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/RecruitmentController.java:34]: Not sure if we want standard users to add recruitment?

File: rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/RecruitmentController.java [hunk 3]
```diff
@@ -20,24 +21,27 @@
...
     @ResponseBody
     @ResponseStatus(value = HttpStatus.CREATED)
     public AddRecruitmentResponse addSystemUser(@RequestBody final AddRecruitmentRequest request) {
         return recruitmentService.saveRecruitment(request);
```
Comment (enthyp) [inline rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/RecruitmentController.java:38]: I know it's not the subject of this PR but we could use the opportunity and fix the names of methods in this class, it's related to recruitments not system users

File: rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/service/UserService.java [hunk 4]
```diff
@@ -100,21 +102,32 @@ public void validateAddSystemUserRequest(final AddSystemUserRequest request) {
...
+    private List<SystemUser> getSystemUsersByType(final UserType type) {
         return userRepository
                 .findSystemUsersByUserType(type)
                 .orElseThrow(() -> new UserNotFoundException("Unable to find users with type: " + type));
     }
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        log.debug(username + " tries to log in");
+        try {
```
Comment (enthyp) [inline rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/service/UserService.java:126]: I think debug messages should be gone (here and in catch block)

File: rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/model/security/UserPrincipal.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+
+import java.util.Collection;
+import java.util.List;
+
+@AllArgsConstructor
+public final class UserPrincipal implements UserDetails {
+    private final SystemUser user;
```
Comment (enthyp) [inline rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/model/security/UserPrincipal.java:14]: Do we need this class or will we need it later? It seems we can just return an instance of org.springframework.security.core.userdetails.User class in UserService.loadUserByUsername method

**Review by enthyp — COMMENTED** [review]
I've added some minor remarks to the code. 

And also:
1. I encountered an issue: I added a user 'root' with password 'pass' to the DB and submitted, got a server error: 
```
rezerwacje-app_1    | java.lang.IllegalArgumentException: There is no PasswordEncoder mapped for the id "null"
```
What am I doing wrong?

2. Merging this PR alone would make working with front-end a pain until NIAK-61 and NIAK-62 are merged since all requests are rejected. 
So I have a suggestion, maybe, once NIAK-55 is fixed, it could be merged into NIAK-61 and then we could leave only NIAK-61 PR open? NIAK-62 is contained in NIAK-61 anyway if I'm correct. Then I shall review NIAK-61 with a shorter delay than I did this branch :)

**Pull request merged**